src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs

```diff
@@ -1212,16 +1213,81 @@ public void AsyncInterfaceMember()
             var source = @"
 interface IInterface
 {
-    async void F(); 
+    async void F();
 }";
             CreateCompilationWithMscorlib45(source).VerifyDiagnostics(
                 // (4,16): error CS0106: The modifier 'async' is not valid for this item
-                //     async void F(); 
+                //     async void F();
                 Diagnostic(ErrorCode.ERR_BadMemberFlag, "F").WithArguments("async"));
         }
 
         [Fact]
-        public void MainCantBeAsync()
+        public void MainCanBeAsyncWithArgs()
+        {
+            var origSource = @"
+using System.Threading.Tasks;
+
+class A
+{
+    async static Task Main(string[] args)
+    {
+        await Task.Factory.StartNew(() => { });
+    }
+}";
+            var sources = new string[] { origSource, origSource.Replace("async ", "").Replace("await", "return") };
+            foreach (var source in sources)
+            {
+                var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7_1));
+                compilation.VerifyDiagnostics();
+                var entry = compilation.GetEntryPoint(CancellationToken.None);
+                Assert.NotNull(entry);
+                Assert.Equal("System.Threading.Tasks.Task A.Main(System.String[] args)", entry.ToTestDisplayString());
+            }
+        }
+
+        [Fact]
+        public void MainCantBeAsyncWithArgs_CSharp7()
+        {
+            var source = @"
+using System.Threading.Tasks;
+
+class A
+{
+    async static Task Main(string[] args)
+    {
+        await Task.Factory.StartNew(() => { });
+    }
+}";
+            var compilation = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7));
+            compilation.VerifyDiagnostics(Diagnostic(ErrorCode.ERR_MainCantBeAsync, "Main").WithArguments("A.Main(string[])").WithLocation(6, 23));
+        }
+
+        [Fact]
+        public void MainCanBeAsync()
```

Comment:
Note: please add a test verifying we don't allow variants where main has a ref return.

``` C#
static ref Task Main() { ... }
```

Context is that we incorrectly allow this for `int` today: https://github.com/dotnet/roslyn/issues/17923